### PR TITLE
TensorGraph layer tutorial using eager execution

### PR DESCRIPTION
**Pull request by nitinprakash96**
This notebook currently demonstrates:
- `Dense` and `Conv1D` layer construction
- How to use `create_tensor()` and `Dense(x)`
- Execution under eager mode

Still to add:
- How to take gradients under eager execution

### CODE REVIEW

**Comment by peastman** [comment]
How about adding more text and providing more context for everything?  What is eager execution mode?  How does it differ from the standard mode?  Why would you want to use it?  Assume the reader is just clicking through the DeepChem tutorials and starts out having no idea what this one is about.

**Comment by nitinprakash96** [comment]
@peastman sure. I'll add more information about eager mode. Perhaps at the start?

**Comment by peastman** [comment]
That's looking a lot better!

**Comment by rbharath** [comment]
@nitinprakash96 Wow! This looks really good! Tensorflow Eager really does look like it can be the API of the future :)

CC @patrickhop: From your TensorFlow eager experience, is there anything else that might be good to add in here?

**Comment by nitinprakash96** [comment]
@rbharath Thanks!
As for the gradient part. I am not really getting what our snippets should be. Could you help me out in that.

**Comment by rbharath** [comment]
Hmm, here's some thoughts for what we might demonstrate for gradients:
```
def double_dense(x):
  return Dense(1)(Dense(1)(x))

grad = tfe.gradients_function(double_dense)

print(double_dense(3.))
print(grad(3.))
```
Feel free to change this around if there's a nicer example we can make.

**Comment by peastman** [comment]
I looked through the code to see what's actually happening, and concluded we've just jumped *way* ahead of ourselves.  This is a nice tutorial about a feature that, for most practical purposes, doesn't exist yet!

Consider this line from the tutorial:

```python
y = Dense(3)(inputs)
```

Here's what that line does.  It creates a dense layer with three outputs.  It initializes the weights to random values, and the biases to 0.  And then it multiplies the input tensor by the random weights, which is nice but totally useless!

For TensorGraph layers to actually be useful in eager mode, they need to store actual weights.  You need to be able to specify those weights when you create them.  You need to be able to run a training loop and update those weights at each step.  And all this needs to happen without a TensorGraph or a tf.Session.  In short, this is going to take a lot of new infrastructure that doesn't exist yet.

**Comment by nitinprakash96** [comment]
@peastman The thing about it multiplying the input weight by random tensors is true and would not make much sense to the user on whats going on. But I think in general it gives an idea how the function works without going through the codebase. And perhaps would help in understanding other tutorials wherever TensorGraph layers are used.

**Comment by rbharath** [comment]
@peastman Good points! It's going to take a lot more work before the DeepChem Eager support is full fledged...

@nitinprakash96 Agree that this tutorial adds in a good chunk of value already! Would you mind adding on a section that discusses the points @peastman raises? We should explain to readers that the tutorial is using random weights for now.

**Comment by peastman** [comment]
I think we should put this on hold.  There's nothing wrong with the tutorial itself—it accurately describes what currently exists.  The problem is that what currently exists just isn't useful.  If we post a tutorial for something that is essentially a non-existing feature, that's only going to confuse people.  Let's first do the work to properly support eager mode as a useful tool.  Once that's done, we can finish up this tutorial.

**Comment by rbharath** [comment]
Took a deeper look into what it would take to train DeepChem models in Eager mode (let's discuss in #1171), and it's going to involve a little work

@nitinprakash96: Will us keeping this PR in hold while we figure out the training issues negatively impact your Google Summer of Code application?

If so, one option might be for us to merge, but mark this as a WIP notebook which we won't upload to the website until the Eager integration is functional.

**Comment by nitinprakash96** [comment]
@rbharath I don't think it should do any harm. It's probably better to keep this open until the issue is solved.

**Comment by peastman** [comment]
I think we can resume work on this now that #1176 and #1191 have been merged.  Here are some thoughts on what to discuss in it.

- I wouldn't mention calling `create_tensor()`.  Just tell them to invoke the layer as a function.  That matches the API of Tensorflow and Keras layers.  Also, we may not always want to support calling `create_tensor()` directly.

- TensorGraph also supports `__call()__`, so you can invoke a model in eager mode exactly the same way you would a single layer.

- You can even mix and match Tensorflow and DeepChem layers.  You can't add Tensorflow layers to a TensorGraph, of course, but if you're just invoking them directly that isn't a problem.  They all take tensors as inputs and return tensors as outputs, so you can take the output from one kind of layer and pass it as input to a different kind of layer.

- If a layer defines variables, they are created the first time the layer is invoked.  They might depend on the sizes of the inputs.  For example, a Dense layer has a matrix of weights of shape (N, M), where M is the number of inputs and N is the number of outputs.  The first time you call it, the layer sets M based on the length of the input you passed to it and creates the weight matrix.  If you later try to pass an input of a different length to that layer, you'll get an error.  This is identical to how Tensorflow layers work.

- Training variables works exactly the same as for Tensorflow layers.  See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/eager/python/g3doc/guide.md for an example.  If `loss()` is a function that takes a batch of training data and returns the loss, you can write

```python
grad = tfe.implicit_gradients(loss)
for batch in generator:
    optimizer.apply_gradients(grad(batch))
```

- Remember that eager mode always defaults to running on the CPU, unlike graph mode that defaults to the GPU!  The tutorial linked above describes how to move the calculations to the GPU.

**Comment by nitinprakash96** [comment]
@peastman I'll update the PR soon and let you know about the progress.

**Comment by rbharath** [comment]
@nitinprakash96 Do you have time to finish this tutorial? Now that we have eager support, this would add a lot of value. If you're too busy, that's fine as well. Just let us know so one of us can complete the tutorial.

**Comment by nitinprakash96** [comment]
Sorry missed this one. On it.

**Comment by nitinprakash96** [comment]
@rbharath I've added some points but still need to address the last 2 points by @peastman. Let me know if this is the right direction.

**Comment by rbharath** [comment]
@nitinprakash96 Looking good! I'd add some lines emphasizing that all these examples are using random intializations and then add on examples of how to train eager models as @peastman suggests above

**Comment by nitinprakash96** [comment]
@peastman Let's say we were training an MNIST model in eager mode, can we invoke the layers as following?

```python
tg = dc.models.TensorGraph(tensorboard=False, model_dir='/tmp/mnist')
feature = Feature(shape=(None, 784))

image = Reshape(shape=(-1, 28, 28, 1))(feature)

conv2d_1 = Conv2D(num_outputs=32)(image)

conv2d_2 = Conv2D(num_outputs=64)(conv2d_1)

flatten = Flatten()(conv2d_2)

dense1 = Dense(out_channels=1024, activation_fn=tf.nn.relu)(flatten)

dense2 = Dense(out_channels=10)(dense1)
```

**Comment by peastman** [comment]
Not quite.  First of all, there's no reason to create a TensorGraph.  You're just invoking the layers directly, not adding them to a TensorGraph.

Next, you wouldn't create a Feature layer.  That's only for use with TensorGraph.  It serves as a placeholder to indicate where the inputs should be injected into the graph.  But in this case, you would simply pass the inputs directly to the relevant layers.

Next, you generally shouldn't construct and invoke a layer at the same time.  Consider this line from your code:

```python
conv2d_1 = Conv2D(num_outputs=32)(image)
```

That line 1) creates a Conv2D layer with random weights, 2) invokes it with `image`, and 3) throws away the Conv2D layer because you don't save any reference to it!  Instead you should create the layer and assign it to a variable.  You will then invoke that layer many times over the course of training so its weights can be gradually optimized.

For an example of this, see the "build a model" section of https://www.tensorflow.org/guide/eager.  It uses Keras layers, but it basically works the same way.

**Comment by rbharath** [comment]
I'm working on overhauling the tutorials right now. Would anyone here mind if I go ahead and merge this tutorial in? There's some good material in here on Eager mode that I can adapt to Keras.

**Comment by nitinprakash96** [comment]
@rbharath No worries. Please do
